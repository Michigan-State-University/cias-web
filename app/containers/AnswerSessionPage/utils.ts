import dayjs from 'dayjs';
import pickBy from 'lodash/pickBy';
import isNil from 'lodash/isNil';

import { TlfbConfigBody } from 'models/Question';
import { CalendarData } from 'models/Tlfb';
import { CamelToSnake } from 'global/types/camelToSnake';
import { fullDayToYearFormatter } from 'utils/formatters';

export const getSessionMapUserPreviewUrl = (userSessionId: string): string => {
  const {
    location: { pathname },
  } = window;
  const previewIndex = pathname.indexOf('preview');
  const sessionUrl = pathname.slice(0, previewIndex - 1);
  return `${sessionUrl}/map?userSessionId=${userSessionId}`;
};

export const getNextSessionUrl = (sessionId: string): string => {
  const {
    location: { pathname },
  } = window;

  return pathname.replace(/\/sessions\/.*/, `/sessions/${sessionId}/fill`);
};

export const getBackToModulesUrl = (): string => {
  const {
    location: { pathname },
  } = window;

  return pathname.replace(/\/sessions\/.*/, '/invite');
};

const generateGridCellId = (prefix: string, index: number) =>
  `${prefix}-${index}`;

export const generateGridAnswerId = (index: number) =>
  `${generateGridCellId('answer', index)}`;
export const generateGridQuestionId = (index: number) =>
  `${generateGridCellId('question', index)}`;

const getOldestDate = (dates: CalendarData) =>
  dayjs.min(
    Object.keys(dates).map((dayString) =>
      dayjs(dayString, fullDayToYearFormatter),
    ),
  );

export const getCalendarMetadata = (
  tlfbConfig: CamelToSnake<TlfbConfigBody>,
  calendarData: CalendarData,
) => {
  const {
    data: [
      {
        payload: { days_count: daysCount },
      },
    ],
  } = tlfbConfig;
  const yesterday = dayjs().subtract(1, 'day');
  const oldestFilledDate = getOldestDate(calendarData);
  const datesWithAnswers = pickBy(calendarData, (date) => !isNil(date.answer));
  const oldestFilledAnswerDate = getOldestDate(datesWithAnswers);
  const oldestAllowedDate = dayjs().subtract(+(daysCount || 1), 'day');
  const isEveryAnswerFilled =
    oldestFilledAnswerDate?.isSame(oldestAllowedDate, 'day') ?? false;

  return {
    yesterday,
    oldestAllowedDate,
    oldestFilledDate,
    oldestFilledAnswerDate,
    isEveryAnswerFilled,
  };
};
