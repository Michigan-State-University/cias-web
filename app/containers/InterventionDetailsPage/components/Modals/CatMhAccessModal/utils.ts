import { Intervention } from 'models/Intervention';

import { ModalUIData } from './types';

export const getTestsLeft = (
  catMhPool: Nullable<number>,
  createdCatMhSessionCount: Nullable<number>,
) => Math.max(0, (catMhPool ?? 0) - (createdCatMhSessionCount ?? 0));

export const mapInterventionToModalData = (
  intervention: Intervention,
): ModalUIData => {
  const {
    catMhPool,
    licenseType,
    catMhApplicationId,
    catMhOrganizationId,
    isAccessRevoked,
    createdCatMhSessionCount,
  } = intervention;

  return {
    testNumber: catMhPool ?? 0,
    testsLeft: getTestsLeft(catMhPool, createdCatMhSessionCount),
    licenseType,
    applicationId: catMhApplicationId ?? '',
    organizationId: catMhOrganizationId ?? undefined,
    isAccessRevoked,
  };
};

export const mapModalDataToIntervention = (
  modalData: ModalUIData,
): Partial<Intervention> => {
  const {
    testNumber,
    licenseType,
    applicationId,
    organizationId,
    isAccessRevoked,
  } = modalData;

  return {
    catMhPool: testNumber ?? 0,
    licenseType,
    catMhApplicationId: applicationId,
    catMhOrganizationId: organizationId,
    isAccessRevoked,
  };
};
