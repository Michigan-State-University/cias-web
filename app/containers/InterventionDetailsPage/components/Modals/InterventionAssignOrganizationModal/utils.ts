import { Organization } from 'models/Organization';

import { SelectOption } from 'components/Select/types';

export const organizationSelectOptionFormatter = ({
  id: value,
  name: label,
}: Organization): SelectOption<string> => ({
  value,
  label,
});
