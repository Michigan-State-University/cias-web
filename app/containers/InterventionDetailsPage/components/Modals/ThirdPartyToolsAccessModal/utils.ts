import { InterventionDto } from 'models/Intervention';

import { ModalUIData } from './types';

export const mapInterventionToModalData = (
  intervention: InterventionDto,
): ModalUIData => {
  const {
    catMhPool,
    licenseType,
    catMhApplicationId,
    catMhOrganizationId,
    isAccessRevoked,
    createdCatMhSessionCount,
    hfhsAccess,
  } = intervention;

  return {
    testNumber: catMhPool ?? 0,
    currentTestNumber: catMhPool - createdCatMhSessionCount ?? 0,
    licenseType,
    applicationId: catMhApplicationId ?? '',
    organizationId: catMhOrganizationId ?? undefined,
    isAccessRevoked,
    hfhsAccess,
  };
};

export const mapModalDataToIntervention = (
  modalData: ModalUIData,
): Partial<InterventionDto> => {
  const {
    testNumber,
    licenseType,
    applicationId,
    organizationId,
    isAccessRevoked,
    hfhsAccess,
  } = modalData;

  return {
    catMhPool: testNumber ?? 0,
    licenseType,
    catMhApplicationId: applicationId,
    catMhOrganizationId: organizationId,
    isAccessRevoked,
    hfhsAccess,
  };
};
