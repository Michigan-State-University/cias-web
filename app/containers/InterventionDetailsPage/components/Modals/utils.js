export const organizationSelectOptionFormatter = ({
  id: value,
  name: label,
}) => ({
  value,
  label,
});
