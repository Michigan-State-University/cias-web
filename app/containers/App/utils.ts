import { matchPath } from 'react-router';

import { PasswordAuthenticatedRoles, Roles } from 'models/User/RolesManager';

import { arraysOverlap } from 'utils/arrayUtils';

import { ALL_ROUTE_PATHS, RoutePath } from 'global/constants';

import { OMIT_FETCH_SELF_DETAILS_PATHS } from './constants';

export const shouldFetchSelfDetailsOnPath = (pathname: string) => {
  const match = matchPath(pathname, { path: ALL_ROUTE_PATHS, exact: true });
  return (
    match && !OMIT_FETCH_SELF_DETAILS_PATHS.includes(match.path as RoutePath)
  );
};

export const shouldFetchSelfDetailsByUserRoles = (userRoles: Roles[]) =>
  arraysOverlap(userRoles, PasswordAuthenticatedRoles);
