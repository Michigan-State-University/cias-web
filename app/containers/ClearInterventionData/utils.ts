import * as Yup from 'yup';
import { nonNegativeIntegerValidationSchema } from 'utils/validators';

export const createClearConversationDataFormValidationSchema = () =>
  Yup.object().shape({
    delay: nonNegativeIntegerValidationSchema,
  });
