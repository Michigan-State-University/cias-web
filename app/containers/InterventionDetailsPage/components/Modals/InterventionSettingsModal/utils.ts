import * as Yup from 'yup';

import { Intervention, InterventionType } from 'models/Intervention';
import { ShortLinkData } from 'models/ShortLink';

import { jsonApiToArray } from 'utils/jsonApiMapper';
import { LanguageSelectOption } from 'utils/formatters';
import {
  requiredValidationSchema,
  unreservedURLCharactersSchema,
} from 'utils/validators';

import {
  InterventionSettingsFormValues,
  GetShortLinksResponse,
  ShortLinksData,
} from './types';

// TODO validate duplicates
export const InterventionSettingsFormValidationSchema = Yup.object().shape({
  links: Yup.array().of(
    Yup.object({
      selected: Yup.boolean(),
      name: Yup.string().when('selected', {
        is: (selected) => selected,
        then: unreservedURLCharactersSchema.concat(requiredValidationSchema),
      }),
      healthClinicId: Yup.string(),
    }),
  ),
});

export const getShortLinksDataParser = (
  data: GetShortLinksResponse,
): ShortLinksData => ({
  shortLinks: jsonApiToArray(data, 'shortLink'),
  healthClinics: jsonApiToArray(
    { data: data.health_clinics },
    'simpleHealthClinic',
  ),
});

export const mapShortLinksToFormValues = (
  shortLinksData: Nullable<ShortLinksData>,
  inOrganization: boolean,
): InterventionSettingsFormValues['links'] => {
  if (!inOrganization) {
    const links = shortLinksData?.shortLinks?.map(({ name }) => ({
      selected: !!name,
      name: name ?? '',
    }));
    return links?.length ? links : [{ selected: false, name: '' }];
  }

  return (
    shortLinksData?.healthClinics?.map(({ id }) => {
      const { name } =
        shortLinksData?.shortLinks?.find(
          ({ healthClinicId }) => healthClinicId === id,
        ) ?? {};
      return {
        selected: !!name,
        name: name ?? '',
        healthClinicId: id,
      };
    }) ?? []
  );
};

export const mapFormValuesToShortLinks = (
  links: InterventionSettingsFormValues['links'],
): ShortLinkData[] =>
  links
    .filter(({ selected, name }) => selected && name)
    .map(({ name, healthClinicId }) => ({ name: name!, healthClinicId }));

export const mapLanguageToInterventionChanges = ({
  value,
  label,
  googleLanguageId,
}: LanguageSelectOption): Pick<
  Intervention,
  'languageCode' | 'languageName' | 'googleLanguageId'
> => ({
  languageCode: value,
  languageName: label,
  googleLanguageId: +googleLanguageId,
});

export const getPlaceholderBase = (
  interventionId: string,
  interventionType: InterventionType,
  firstSessionId: Nullable<string>,
) => {
  const base = `${process.env.WEB_URL}/interventions/${interventionId}`;

  switch (interventionType) {
    case InterventionType.DEFAULT:
      if (!firstSessionId) return '';
      return `${base}/sessions/${firstSessionId}/fill`;
    default:
      return `${base}/invite`;
  }
};
