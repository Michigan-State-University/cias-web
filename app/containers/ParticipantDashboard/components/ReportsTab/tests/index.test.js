/**
 *
 * Tests for ReportsTab
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { DEFAULT_LOCALE } from 'i18n';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';

import { Provider } from 'react-redux';

import { createTestStore } from 'utils/testUtils/storeUtils';
import { createReport } from 'utils/reducerCreators';
import ReportsPage from '../index';

describe('<ReportsPage />', () => {
  let store;
  beforeAll(() => {
    store = createTestStore({
      dashboard: {
        reports: [createReport(1)],
        reportsSize: 1,
        loaders: {
          fetchReportsLoading: false,
        },
        errors: {},
      },
    });
  });

  it('Should render and match the snapshot and to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');

    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <MemoryRouter>
            <ReportsPage />
          </MemoryRouter>
        </IntlProvider>
      </Provider>,
    );

    expect(spy).not.toHaveBeenCalled();
    expect(firstChild).toMatchSnapshot();
  });
});
