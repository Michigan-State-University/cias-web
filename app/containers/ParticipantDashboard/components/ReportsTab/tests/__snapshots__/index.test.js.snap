// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportsPage /> Should render and match the snapshot and to not log errors in console 1`] = `
<div
  class="styled__StyledContainer-sc-6h8fd4-0 fziHXq"
  style="box-sizing: border-box; position: relative; margin-left: auto; margin-right: auto; padding-left: 15px; padding-right: 15px; max-width: 960px;"
>
  <h1
    class="H1-sc-1k7m23t-0 gkCKau"
  >
    My Reports
  </h1>
  <div
    class="Column-sc-4zswzq-0 jCuscc"
  >
    <div
      class="Row-y1cs0u-0 jdfDUM"
      width="100%"
    >
      <div
        class="Box-sc-1vhutan-0 styled__StyledBox-onmxmg-0 bqTbKg erHOle"
      >
        <div
          height="100%"
          style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: center; justify-content: flex-start;"
        >
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 75%; flex-basis: 75%; flex-grow: 0; flex-shrink: 0; max-width: 75%; margin-left: 0%;"
          >
            <div
              class="Row-y1cs0u-0 ecllgQ"
            >
              <img
                src="IMAGE_MOCK"
              />
              <p
                class="Text-e4p0w3-0 fxvuEz"
                font-size="18"
                font-weight="bold"
              >
                Report: 1. part - Results
              </p>
            </div>
          </div>
          <div
            style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 25%; flex-basis: 25%; flex-grow: 0; flex-shrink: 0; max-width: 25%; margin-left: 0%;"
          >
            <div
              class="Row-y1cs0u-0 styled__DownloadRow-onmxmg-1 ecllgQ NPQnz"
            >
              <a
                class="StyledLink-eun179-0 jlCCcr"
                href="/file.pdf"
                target="_blank"
              >
                <p
                  class="Text-e4p0w3-0 fOFYCf"
                  color="#C866EA"
                  font-size="13"
                  font-weight="bold"
                >
                  <span>
                    Download
                  </span>
                </p>
                <img
                  src="IMAGE_MOCK"
                />
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Row-y1cs0u-0 cAHLWQ"
    display="flex"
    width="100%"
  >
    <div
      class="Box-sc-1vhutan-0 cfDdZL"
      cursor="pointer"
      display="flex"
    >
      <p
        class="Text-e4p0w3-0 eRNSEk"
        color="#FFFFFF"
      >
        1
      </p>
    </div>
  </div>
  <span
    style="display: table; clear: both;"
  />
</div>
`;
